lib/spack/spack/solver/asp.py

```diff
@@ -116,8 +116,32 @@ def getter(node):
     version_origin_fields,
 )(**{name: i for i, name in enumerate(version_origin_fields)})
 
-#: Named tuple to contain information on declared versions
-DeclaredVersion = collections.namedtuple("DeclaredVersion", ["version", "idx", "origin"])
+
+class DeclaredVersion(object):
+    def __init__(self, version, idx, origin):
+        if not isinstance(version, spack.version.VersionBase):
+            raise ValueError("Unexpected type for declared version: {0}".format(type(version)))
+        self.version = version
+        self.idx = idx
+        self.origin = origin
+
+    def __eq__(self, other):
+        if not isinstance(other, DeclaredVersion):
+            return False
+        n_gitversions = sum(
+            1 for x in [self.version, other.version] if isinstance(x, spack.version.GitVersion)
+        )
+        if n_gitversions == 1:
```

Comment:
I find this more intuitive and readable:

```suggestion
        is_git = lambda v: isinstance(v, spack.version.GitVersion)
        if is_git(self.version) != is_git(other.version):
```